### DISCUSSION TITLE
memory error

### DISCUSSION THREAD

**Discussion by eric-yu-zhu** [CederGroupHub/chgnet#89] [General]
Dear Bowen,

Thank you so much for developing and sharing chgnet! I am trying to apply it to a research project. 

The installation of chgnet was very smooth and I am able to run the example of mp-18767-LiMnO2.cif.

However, I ran into an error when running chgnet with my own system which contains 508 atoms. The input files and error message are here https://drive.google.com/drive/folders/1ZUaZ1HULmdIL7-89JPFI6rIa6yRuqyTu?usp=sharing

(1) The problem is caused by memory explosion (run01_mof_md.py, errmsg.txt)
   File "pymatgen/optimization/neighbors.pyx", line 226, in pymatgen.optimization.neighbors.find_points_in_spheres   
   MemoryError: A realloc of memory of failed!
					  
(2) I tried to run the job on a node with 180 GB memory, same error appeared when only 10% memory was being used.

(3) I tried to do the energy/forces/stress prediction, no error (run02_mof_pred.py). It seems the error is related to md?

Could you please help me fix the problem?

I also noticed that the output says: "CHGNet will run on cpu". How do I run chgnet on GPU? Any documents or examples?

Best,
Eric

**Comment by eric-yu-zhu** [comment]
1. The memory issue has been resolved. The problem was caused by fractional coordinates. In the xyz file, the atom positions are Cartesian coordinates whereas pymatgen Structure reads fractional coordinates. Below is the code snippet for loading structures from an xyz file. Hope it helps others.

from pymatgen.core import Structure
from ase.io import read, write
xyz_file = 'MOF5_H2O.xyz'
unitcell = [[26.070007, 0, 0],  [0, 26.070007, 0], [0, 0, 26.070007]]   #Ang
atoms = read(xyz_file, format='xyz')
atoms.set_cell(unitcell)
structure = Structure(atoms.get_cell(), atoms.get_chemical_symbols(), atoms.get_scaled_positions())

2. I still want to know how to run chgnet on GPU, the model inference using CPU is too slow (around 1s per md step). Help needed.

**Comment by bowen-bd** [comment]
1. did you try reading xyz file [using pymatgen](https://pymatgen.org/pymatgen.io.html#pymatgen.io.xyz.XYZ)? 
2. The MD will default to GPU if available, it can also be manually set by `md = MolecularDynamics(use_device='cuda')`. If you're unable to use cuda please check your pytorch installation.

**Comment by eric-yu-zhu** [comment]
1. Yes, pymatgen might be an alternative solution. The key is to use fractional coordinates rather than Cartesian ones.

2. After installing pytorch in the virtual environment, the gpu version of chgnet works. 

Thanks a lot!
